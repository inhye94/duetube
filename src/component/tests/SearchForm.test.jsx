import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { Route } from "react-router-dom";
import { withRouter } from "../../tests/utils";
import SearchBar from "../SearchForm";

// 테스트용 검색어와 이벤트 핸들러
const keyword = "테스트 검색어";
const blurEvent = jest.fn();

// react-router-dom의 useNavigate를 mocking
const mockedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));

describe("SearchForm", () => {
  describe("초기 렌더링", () => {
    it("keyword가 없을 때", () => {
      const { asFragment } = render(
        withRouter(
          <Route path="/" element={<SearchBar blurEvent={blurEvent} />} />
        )
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it("keyword가 있을 때, input 값 업데이트", () => {
      render(
        withRouter(
          <Route
            path="/search/:keyword"
            element={<SearchBar blurEvent={blurEvent} />}
          />,
          `/search/${keyword}`
        )
      );

      // getByDisplayValue = input의 value가 특정 값일 때 찾기
      expect(screen.getByDisplayValue(keyword)).toBeInTheDocument();
    });
  });

  describe("Submit이 되지 않는 경우", () => {
    beforeEach(() => {
      mockedNavigate.mockClear();
    });

    it("검색어가 빈 값", () => {
      const { getByRole } = render(
        withRouter(
          <>
            <Route path="/" element={<SearchBar blurEvent={blurEvent} />} />
          </>
        )
      );

      const button = getByRole("button", { name: "검색" });
      fireEvent.submit(button);

      expect(mockedNavigate).toHaveBeenCalledTimes(0);
    });

    it("검색어가 이전 값과 동일", () => {
      const { getByRole } = render(
        withRouter(
          <Route
            path={`/search/${keyword}`}
            element={<SearchBar blurEvent={blurEvent} />}
          />,
          `/search/${keyword}`
        )
      );

      const submitButton = getByRole("button", { name: "검색" });
      fireEvent.click(submitButton);

      expect(mockedNavigate).toHaveBeenCalledTimes(0);
    });
  });

  describe("Submit이 되고, 페이지 이동", () => {
    let searchButton, searchInput, searchForm;

    beforeEach(() => {
      mockedNavigate.mockClear();

      const { getByRole, getByPlaceholderText, container } = render(
        withRouter(
          <>
            <Route path="/" element={<SearchBar blurEvent={blurEvent} />} />
            <Route path={`/search/${keyword}`} element={<p>{keyword}</p>} />
          </>,
          "/"
        )
      );

      searchForm = container.firstChild;
      searchButton = getByRole("button", { name: "검색" });
      searchInput = getByPlaceholderText("검색");
    });

    it("다른 검색어 입력하고 검색 버튼을 클릭", () => {
      // useNavigate가 mocking 되어있어서, navigate가 호출되는지 & navigate의 어떤 인자가 넘어가는지만 검증하면 됨
      fireEvent.change(searchInput, { target: { value: keyword } });
      fireEvent.click(searchButton);

      expect(mockedNavigate).toHaveBeenCalledTimes(1);
      expect(mockedNavigate).toHaveBeenCalledWith(`/search/${keyword}`);
    });

    it("다른 검색어 입력하고 엔터", () => {
      fireEvent.change(searchInput, { target: { value: keyword } });
      fireEvent.submit(searchForm);

      expect(mockedNavigate).toHaveBeenCalledTimes(1);
      expect(mockedNavigate).toHaveBeenCalledWith(`/search/${keyword}`);
    });
  });
});
