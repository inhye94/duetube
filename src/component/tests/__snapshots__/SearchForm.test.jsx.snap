// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm 초기 렌더링 keyword가 없을 때 1`] = `
<DocumentFragment>
  <form
    class="flex flex-nowrap items-center w-full h-[40px] *:h-full *: *:font-medium *:dark:text-slate-100 *:transition-colors *:duration-300"
  >
    <input
      autocomplete="off"
      class="grow w-[50px] px-[16px] rounded-l-md text-[16px] text-purple-600 bg-slate-100 caret-purple-600 placeholder:text-slate-400 dark:bg-slate-600"
      placeholder="검색"
      type="text"
      value=""
    />
    <button
      aria-label="검색"
      class="shrink-0 px-8 rounded-r-md text-[24px] bg-slate-200 hover:bg-slate-300 dark:bg-slate-500 dark:hover:bg-slate-400"
      title="검색 버튼을 눌러주세요"
      type="submit"
    >
      <svg
        aria-label="검색"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M456.69 421.39L362.6 327.3a173.81 173.81 0 0034.84-104.58C397.44 126.38 319.06 48 222.72 48S48 126.38 48 222.72s78.38 174.72 174.72 174.72A173.81 173.81 0 00327.3 362.6l94.09 94.09a25 25 0 0035.3-35.3zM97.92 222.72a124.8 124.8 0 11124.8 124.8 124.95 124.95 0 01-124.8-124.8z"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;
